specs/exponentiation-circuit.md

```diff
@@ -0,0 +1,254 @@
+* `exponentiation`: RLC-encoded 32-byte word, pushed to the stack
+* `exponent_is_zero`: `IsZeroGadget` to check whether or not `exponent == 0`
+* `exponent_is_one`: `IsEqualGadget` to check whether or not `exponent == 1`
+* `single_step`: A boolean to indicate whether or not there will be a single step in the exponentiation trace. There is a single step if `exponent == 2`, which means we have `base * base == base^2 (mod 2^256)` as the only step. We use this to determine the `is_first` and `is_last` column values, while doing lookups to the exponentiation table.
+* `exponent_byte_size`: `ByteSizeGadget` to check the byte-size of `exponent`, i.e. the minimum number of bytes that are required to represent `exponent`. In python, the code for byte size would be:
+```
+byte_size_value = (value.bits() + 7) // 8
+```
+
+### Constraints
+
+* If `exponent == 0` then `exponentiation == 1`
+* If `exponent == 1` then `exponentiation == base`
+* If `exponent == 2` then:
+    * Do lookup to the `ExpTable`:
+    ```
+    (
+        is_last=1,
+        base_limbs,
+        exponent_lo_hi,
+        exponentiation_lo_hi
+    )
+    ```
+* If `exponent > 2` then:
+    * Do lookup to the `ExpTable`:
+    ```
+    (
+        is_last=0,
+        base_limbs,
+        exponent_lo_hi,
+        exponentiation_lo_hi,
+    )
+    ```
+    * Do lookup to the `ExpTable`:
+    ```
+    (
+        is_last=1,
+        base_limbs,
+        exponent_lo_hi=[2, 0],
+        exponentiation_lo_hi=[base_sq_lo, base_sq_hi],
+    )
+    ```
+    where `base_sq_lo` and `base_sq_hi` are the 128-bit low-high parts of `base ^ 2 (mod 2^256)`.
+* If `exponent == 0` then `exponent_byte_size == 0`
+* Gas cost `gas == constant_gas_cost(EXP) +  dynamic_gas_cost` where `dynamic_gas_cost == 50 * exponent_byte_size`.
+
+--------------------------------------------------------
+
+## Byte Size Gadget
+
+We wish to calculate an expression that represents the byte size of a 32-bytes RLC-encoded word.
+
+We make use of an array of 33 cells, called `most_significant_nonzero_byte_index`. In this array, exactly one cell will be turned on (set to `1`) while the rest are all `0`. The turned on cell indicates the index of the most significant non-zero byte in the 32-bytes word.
+
+If this index is such that `index > 0`, this means that the byte size is non-zero. Which also means that there exists an inverse of the byte at that index in the 32-bytes word.
+
+We make use of a cell called `most_significant_nonzero_byte_inverse` to store the inverse of the byte at the said index.
+
+### Constraints
+
+* Exactly one cell is turned on from the array `most_significant_nonzero_byte_index`.
+* For instance, if the above turned on index is `index'`, then we have:
+```
+sum(word_bytes[index'..32]) == 0
+```
+which basically means that, all the other bytes from `index'` must be `0` (since in the little-endian representation of word, bytes beyong the byte-size will be `0`).
```

Comment:
```suggestion
which basically means that, all the other bytes from `index'` must be `0` (since in the little-endian representation of word, bytes beyond the byte-size will be `0`).
```